src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,485 @@
 
-import Algebra.Graph.Relation
-import Algebra.Graph.Relation.InternalDerived
+import qualified Algebra.Graph.Relation as R
+import qualified Algebra.Graph.Relation.Internal as RI hiding (empty, vertex, overlay, connect, referredToVertexSet)
+import Algebra.Graph.Relation.Symmetric.Internal
 
 import qualified Data.Set as Set
+import Data.Tree 
+import qualified Data.Tree as Tree
+import Data.Tuple (swap)
 
 -- | Construct a symmetric relation from a 'Relation'.
--- Complexity: /O(1)/ time.
-fromRelation :: Relation a -> SymmetricRelation a
-fromRelation = SymmetricRelation
+-- Complexity: /O(m*log(m))/ time.
+--
+-- @
+-- fromRelation ('Algebra.Graph.Relation.edge' 1 2) == 'edge 1 2'
+-- fromRelation . 'toRelation'                      == id
+-- 'vertexCount' . fromRelation                     == 'Algebra.Graph.Relation.vertexCount'
+-- (* 2) . 'edgeCount' . fromRelation               >= 'Algebra.Graph.Relation.edgeCount'
+-- @
+fromRelation :: Ord a => R.Relation a -> SymmetricRelation a
+fromRelation = SymmetricRelation . R.symmetricClosure
 
 -- | Extract the underlying relation.
--- Complexity: /O(m*log(m))/ time.
-toRelation :: Ord a => SymmetricRelation a -> Relation a
-toRelation = symmetricClosure . fromSymmetric
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- toRelation ('edge' 1 2)                           == 'Algebra.Graph.Relation.edges' [(1,2), (2,1)]
+-- 'Algebra.Graph.Relation.vertexCount' . toRelation == 'vertexCount'
+-- 'Algebra.Graph.Relation.edgeCount' . toRelation   <= (* 2) . 'edgeCount'
+-- @
+toRelation :: SymmetricRelation a -> R.Relation a
+toRelation = fromSymmetric
+
+-- | Construct the graph comprising /a single edge/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- edge x y                 == edges [(x,y), (y,x)]
+-- edge x y                 == connect (vertex x) (vertex y)
+-- edge x y                 == edge y x
+-- 'hasEdge' x y (edge x y) == True
+-- 'edgeCount'   (edge x y) == 1
+-- 'vertexCount' (edge 1 1) == 1
+-- 'vertexCount' (edge 1 2) == 2
+-- @
+edge :: Ord a => a -> a -> SymmetricRelation a
+edge x y = SymmetricRelation $ RI.Relation (Set.fromList [x, y]) (Set.fromList [(x, y), (y,x)]) 
+
+-- | Construct the graph comprising a given list of isolated vertices.
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is the length
+-- of the given list.
+--
+-- @
+-- vertices []              == 'empty'
+-- vertices [x]             == 'vertex' x
+-- 'hasVertex' x . vertices == 'elem' x
+-- 'vertexCount' . vertices == 'length' . 'Data.List.nub'
+-- 'vertexSet'   . vertices == Set.'Set.fromList'
+-- @
+vertices :: Ord a => [a] -> SymmetricRelation a
+vertices = SymmetricRelation . R.vertices
+
+-- | Construct the graph from a list of edges.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- edges []            == 'empty'
+-- edges [(x,y)]       == 'edge x y'
+-- 'edgeCount' . edges == 'length' . 'Data.List.nub' . 'uncurry' 'Data.List.union' . '<map swap, id>'
```

Comment:
What does `<map swap, id>` mean?

Reply:
I mean its the split operator that takes an input and returns a pair that runs one function on each component. It's equivalent to this `(,) <$> map swap <*> id` but I think this is harder to read. Is there a better notation?

Reply:
I don't use any non-Haskell notation in comments, so you'll need to find a way to express this property in Haskell!

Reply:
I went with `(,) <$> map swap <*> id`

Reply:
Examples are only useful if readers can easily understand them. I think this example is just too complex and therefore not very useful. Try to simplify it.

Reply:
Okay! Can I add something like `where` or `let` to declare an auxiliary function to be more specific?

Reply:
I would prefer to stick to simple one-line properties. Don't spend too much time on this -- if you can't figure out a simpler example, simply remove it.

I'd probably add something like `edges [(x, y), (y, x)] == edge x y` instead.

Reply:
I'll follow your suggestion.